linux/GPU.c

```diff
@@ -0,0 +1,272 @@
+};
+
+enum section_state {
+   SECST_UNKNOWN,
+   SECST_DUPLICATE,
+   SECST_NEW,
+};
+
+static bool null_str_eq(const char* s1, const char* s2) {
+   if (s1 == NULL && s2 == NULL)
+      return true;
+
+   if (s1 && s2)
+      return String_eq(s1, s2);
+
+   return false;
+}
+
+static bool is_duplicate_client(const struct client_id* parsed, unsigned long long int id, const char* pdev) {
+   for (; parsed; parsed = parsed->next) {
+      if (id == parsed->id && null_str_eq(pdev, parsed->pdev)) {
+         return true;
+      }
+   }
+
+   return false;
+}
+
+static void update_system(LinuxProcessList* lpl, unsigned long long int time, char* engine) {
+   Machine* host = lpl->super.host;
+   LinuxMachine* lhost = (LinuxMachine*) host;
+   GPUEngineData** engineData = &lhost->gpuEngineData;
+
+   while (*engineData) {
+      if (String_eq((*engineData)->key, engine))
+         break;
+
+      engineData = &((*engineData)->next);
+   }
+
+   if (!*engineData) {
+      GPUEngineData* newData;
+
+      newData = xMalloc(sizeof(*newData));
+      *newData = (GPUEngineData) {
+         .prevTime = 0,
+         .curTime  = 0,
+         .key      = engine,
+         .next     = NULL,
+      };
+
+      *engineData = newData;
+   } else
+      free(engine);
+
+   (*engineData)->curTime += time;
+   lhost->curGpuTime += time;
+}
+
+/*
+ * Documentation reference:
+ * https://www.kernel.org/doc/html/latest/gpu/drm-usage-stats.html
+ */
+void GPU_readProcessData(LinuxProcessList* lpl, LinuxProcess* lp, openat_arg_t procFd) {
+   int fdinfoFd = -1;
+   DIR* fdinfoDir = NULL;
+   struct client_id* parsed_ids = NULL;
+   unsigned long long int new_gpu_time = 0;
+
+   fdinfoFd = Compat_openat(procFd, "fdinfo", O_RDONLY | O_NOFOLLOW | O_DIRECTORY | O_CLOEXEC);
+   if (fdinfoFd == -1)
+     goto out;
+
+   fdinfoDir = fdopendir(fdinfoFd);
+   if (!fdinfoDir)
+      goto out;
+   fdinfoFd = -1;
+
+#ifndef HAVE_OPENAT
+   char fdinfoPathBuf[32];
+   xSnprintf(fdinfoPathBuf, sizeof(fdinfoPathBuf), "/proc/%u/fdinfo", lp->super.pid);
+#endif
+
+   while (true) {
+      const struct dirent* entry;
+      const char* ename;
+      FILE* fp;
+      char* pdev = NULL;
+      unsigned long long int client_id = INVALID_CLIENT_ID;
+      enum section_state sstate = SECST_UNKNOWN;
+      char buf[256];
+
+      entry = readdir(fdinfoDir);
+      if (!entry)
+         break;
+      ename = entry->d_name;
+
+      if (String_eq(ename, ".") ||
+          String_eq(ename, "..") ||
+          String_eq(ename, "0") ||
+          String_eq(ename, "1") ||
+          String_eq(ename, "2"))
+         continue;
+
+#ifdef HAVE_OPENAT
+      fp = Compat_fopenat(dirfd(fdinfoDir), ename, "r");
+#else
+      fp = Compat_fopenat(fdinfoPathBuf, ename, "r");
+#endif
+      if (!fp)
+         continue;
+
+      while (fgets(buf, sizeof(buf), fp)) {
+         if (!String_startsWith(buf, "drm-"))
+            continue;
+
+         if (String_startsWith(buf, "drm-client-id:")) {
```

Comment:
A chain of `String_startsWith` conditionals doesn't look efficient.
Since the token strings you are matching all begin with "drm-", how about using a switch statement?

```c
switch (buf[strlen("drm-")]) {
case 'c':
   if (String_startsWith(buf, "drm-client-id:")) {
      ...
   }
   break;

case 'p':
   if (String_startsWith(buf, "drm-pdev:")) {
      ...
   }
   break;

case 'e':
   if (String_startsWith(buf, "drm-engine:")) {
      ...
   }
   break;
}
```